OSSMDOC-381: Federation configure export/import. (#35973)
openshift/openshift-docs: modules/ossm-federation-create-import.adoc

////
This module included in the following assemblies:
* service_mesh/v2x/ossm-federation.adoc
////

[id="ossm-federation-create-import_{context}"]
= Creating an ImportedServiceSet

You create an `ImportServiceSet` resource to explicitly declare the services that you want to import into your mesh.

Services are imported as `<exported-name>.<exported-namespace>.svc.<peer-name>.remote` which is a "hidden" service, visible only within the egress gateway namespace, along with the "real" service accessible to the mesh ``<export-name>.<export-namespace>.svc.<local-federation-name>.local`. You can treat the local import just like any other service in the mesh. It automatically routes through the egress gateway, where it is redirected to the exported service's remote name.

> .Prerequisites

Reviewer: Add a line break below for this heading to render correctly.

* The cluster and `ServiceMeshControlPlane` have been configured for mesh federation.
* An account with the `cluster-admin` role.

[NOTE]
====
You can configure services for import even if they haven't been exported yet.  When a service that matches the value specified in the ImportServiceSet is deployed and exported, it will be automatically imported.
====

////
.Procedure from the Console
This is conjecture about what the flow might look like.

> Follow this procedure to create a `ImportServiceSet` with the web console. This example shows the green-mesh importing the ratings service that was exported by the red-mesh.

Reviewer: ```suggestion
Follow this procedure to create an `ImportServiceSet` with the web console. This example shows the green-mesh importing the ratings service that was exported by the red-mesh.
```

. Log in to the {product-title} web console as a user with the cluster-admin role.
. Navigate to *Operators* → *Installed Operators*.
. Click the *Project* menu and select the project where you installed the control plane for the mesh you want to import services into. For example, `green-mesh-system`.
. Click the {ProductName} Operator, then click *Istio Service Mesh ImportServiceSet*.
. On the *Istio Service Mesh ImportServiceSet* tab, click *Create ImportServiceSet*.
. On the *Create ImportServiceSet* page, click *YAML* to modify your configuration.
. Modify the default configuration with values for your import.
. Click *Create*. The Operator creates the import the based on your configuration parameters.
. To verify the `ImportServiceSet` resource was created, click the *Istio Service Mesh ImportServiceSet* tab.

> .. Click the name of the new `ImportServiceSet`, for example, `import-from-red-mesh`.

Reviewer: ```suggestion
.. Click the name of the new `ImportServiceSet`; for example, `import-from-red-mesh`.
```

.. Click the *Resources* tab to see the `ImportServiceSet` resource the Operator created and configured.
////

> .Procedure from the CLI

Reviewer: Add a line break below for this heading to render correctly.

//NEED TO TEST THIS

> Follow this procedure to create a `ImportServiceSet` from the command line.

Reviewer: ```suggestion
Follow this procedure to create an `ImportServiceSet` from the command line.
```

. Log in to the {product-title} CLI as a user with the `cluster-admin` role. Enter the following command. Then, enter your username and password when prompted.
+
[source,terminal]
----
$ oc login --username=<NAMEOFUSER> <API token> https://{HOSTNAME}:6443
----
+

> . Change to the project where you installed the control plane, for example, `green-mesh-system`.

Reviewer: ```suggestion
. Change to the project where you installed the control plane; for example, `green-mesh-system`.
```

+
[source,terminal]
----
$ oc project green-mesh-system
----
+

> . Create a `ImportServiceSet` file based on the following example where `green-mesh` is importing services previously exported by `red-mesh`.

Reviewer: ```suggestion
. Create an `ImportServiceSet` file based on the following example where `green-mesh` is importing services previously exported by `red-mesh`.
```

+

> .Example ImportServiceSet resource from red-mesh to green-mesh

Reviewer: Add a line break below for this heading to render correctly.

////
This module included in the following assemblies:
* service_mesh/v2x/ossm-federation.adoc
////

[id="ossm-federation-create-import_{context}"]
= Creating an ImportedServiceSet

You create an `ImportServiceSet` resource to explicitly declare the services that you want to import into your mesh.

> Services are imported as `<exported-name>.<exported-namespace>.svc.<peer-name>.remote` which is a "hidden" service, visible only within the egress gateway namespace, along with the "real" service accessible to the mesh ``<export-name>.<export-namespace>.svc.<local-federation-name>.local`. You can treat the local import just like any other service in the mesh. It automatically routes through the egress gateway, where it is redirected to the exported service's remote name.

Reviewer 2: ```suggestion
Services are imported with the name `<exported-name>.<exported-namespace>.svc.<ServiceMeshPeer.name>.remote` which is a "hidden" service, visible only within the egress gateway namespace and is associated with the exported service's hostname (link to "ossm-federation-create-export_{context}"). The service will be available locally as `<export-name>.<export-namespace>.<domainSuffix>`, where `domainSuffix` is `svc.<ServiceMeshPeer.name>-imports.local` by default, unless `importAsLocal` is `true`, in which case `domainSuffix` is `svc.cluster.local`, otherwise, the domain suffix in the import rule will be applied.  You can treat the local import just like any other service in the mesh. It automatically routes through the egress gateway, where it is redirected to the exported service's remote name.
```

. To verify the `ImportServiceSet` resource was created, click the *Istio Service Mesh ImportServiceSet* tab.
.. Click the name of the new `ImportServiceSet`; for example, `import-from-red-mesh`.
.. Click the *Resources* tab to see the `ImportServiceSet` resource the Operator created and configured.
////

.Procedure from the CLI

//NEED TO TEST THIS
Follow this procedure to create an `ImportServiceSet` from the command line.

. Log in to the {product-title} CLI as a user with the `cluster-admin` role. Enter the following command. Then, enter your username and password when prompted.
+
[source,terminal]
----
$ oc login --username=<NAMEOFUSER> <API token> https://{HOSTNAME}:6443
----
+
. Change to the project where you installed the control plane; for example, `green-mesh-system`.
+
[source,terminal]
----
$ oc project green-mesh-system
----
+
. Create an `ImportServiceSet` file based on the following example where `green-mesh` is importing services previously exported by `red-mesh`.
+
.Example ImportServiceSet resource from red-mesh to green-mesh
[source,yaml]
----
kind: ImportedServiceSet
apiVersion: federation.maistra.io/v1
metadata:

>   name: import-from-red-mesh

Reviewer 2: ```suggestion
  name: red-mesh
```
Needs to match the name of the ServiceMeshPeer resource.
